Subframe navigation is failing in extension popovers

elijahsawyers opened the pull request:
#### fd6c3c9d46a0d20f4f13125568295e1b7fe017fd
<pre>
Subframe navigation is failing in extension popovers
<a href="https://bugs.webkit.org/show_bug.cgi?id=272103">https://bugs.webkit.org/show_bug.cgi?id=272103</a>
<a href="https://rdar.apple.com/125800143">rdar://125800143</a>

Reviewed by Brian Weinstein and Timothy Hatcher.

This patch fixes a bug where subframe navigation is blocked in extension popovers.
We have to loosen the constraints of isUnsupportedWebExtensionNavigation to allow
subframe navigation.

Additionally, this patch adds a new test to validate the fix.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::isUnsupportedWebExtensionNavigation):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIAction, SubframeNavigation)):

Canonical link: <a href="https://commits.webkit.org/277046@main">https://commits.webkit.org/277046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d1dce9a1650f427110b709fe6f091fee5821ca6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49139 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42504 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48770 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23082 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37915 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47041 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40065 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19164 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41171 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4508 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41547 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50974 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17922 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45158 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44095 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6495 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22463 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -514,6 +514,10 @@ static void tryInterceptNavigation(Ref<API::NavigationAction>&& navigationAction
 {
+    bool subframeNavigation = navigationAction.targetFrame() && !navigationAction.targetFrame()->isMainFrame();
+    if (subframeNavigation)
+        return false;
```
I think it might be slightly easier just to pass `subframeNavigation` from both of the methods that call this, instead of computing it again.

xeenon replied:
I think I prefer this, so it is all enclosed. These are inline methods anyway, so it is fast. One of the callosities would need to capture `subframeNavigation` in the lambda too.

b-weinstein replied:
Even for a bool? Interesting!

I retract my comment. I thought if the way we compute `subframeNavigation` ever changed, it would be better to pass it, but that probably won't happen.

elijahsawyers committed:
Subframe navigation is failing in extension popovers
https://bugs.webkit.org/show_bug.cgi?id=272103
rdar://125800143

Reviewed by Brian Weinstein and Timothy Hatcher.

This patch fixes a bug where subframe navigation is blocked in extension popovers.
We have to loosen the constraints of isUnsupportedWebExtensionNavigation to allow
subframe navigation.

Additionally, this patch adds a new test to validate the fix.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::isUnsupportedWebExtensionNavigation):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIAction, SubframeNavigation)):

Canonical link: https://commits.webkit.org/277046@main
